File: sol/solution/0001-0100/0015/Solution.swift
class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        let nums = nums.sorted()
        var res = [[Int]]()
        for i in 0..<nums.count - 2 {
            if i > 0 && nums[i] == nums[i - 1] { continue }
            var l = i + 1, r = nums.count - 1
            while l < r {
                let sum = nums[i] + nums[l] + nums[r]
                if sum < 0 {
                    l += 1
                } else if sum > 0 {
                    r -= 1
                } else {
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r && nums[l] == nums[l + 1] { l += 1 }
                    while l < r && nums[r] == nums[r - 1] { r -= 1 }
                    l += 1
                    r -= 1
                }
            }
        }
        return res
    }
}

File: sol/solution/0001-0100/0015/Solution.php
class Solution {
    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function threeSum($nums) {
        sort($nums);
        $res = [];
        $n = count($nums);
        for ($i = 0; $i < $n - 2; $i++) {
            if ($i > 0 && $nums[$i] == $nums[$i - 1]) continue;
            $l = $i + 1;
            $r = $n - 1;
            while ($l < $r) {
                $sum = $nums[$i] + $nums[$l] + $nums[$r];
                if ($sum < 0) $l++;
                else if ($sum > 0) $r--;
                else {
                    $res[] = [$nums[$i], $nums[$l], $nums[$r]];
                    while ($l < $r && $nums[$l] == $nums[$l + 1]) $l++;
                    while ($l < $r && $nums[$r] == $nums[$r - 1]) $r--;
                    $l++; $r--;
                }
            }
        }
        return $res;
    }
}
